src/lib/game-saving/github.ts

```diff
@@ -0,0 +1,392 @@
+const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+async function handleResponse(response: Response): Promise<any> {
+	if (!response.ok) {
+		const errorData = await response.json().catch(() => ({}));
+		const errorMessage = `GitHub API Error (${response.status}): ${
+			response.statusText
+		} - ${JSON.stringify(errorData)}`;
+		console.error(errorMessage);
+		throw new Error(errorMessage);
+	}
+	return response.json().catch(() => ({}));
+}
+
+async function fetchWithRetry(
+	url: string,
+	options: RequestInit,
+	retries: number = 3,
+	delayMs: number = 1000
+): Promise<Response> {
+	let response: Response;
+	for (let attempt = 0; attempt < retries; attempt++) {
+		response = await fetch(url, options);
+		if (response.ok) return response;
+		if (attempt < retries - 1) {
+			console.warn(
+				`Retrying GitHub API request (${attempt + 1}/${retries})`
+			);
+			await delay(delayMs * (attempt + 1));
+		}
+	}
+	throw new Error("Max retries reached, request failed.");
+}
+
+function getAuthHeaders(accessToken: string): HeadersInit {
+	return {
+		Authorization: `Bearer ${accessToken}`,
+		Accept: "application/json",
+	};
+}
+
+function getJsonHeaders(accessToken?: string): HeadersInit {
+	const headers: HeadersInit = {
+		"Content-Type": "application/json",
+		Accept: "application/json",
+	};
+	if (accessToken) {
+		headers.Authorization = `Bearer ${accessToken}`;
+	}
+	return headers;
+}
+
+export async function fetchGitHubAccessToken(
+	code: string
+): Promise<string | null> {
+	try {
+		const response = await fetch(
+			`https://github.com/login/oauth/access_token`,
+			{
+				method: "POST",
+				headers: {
+					"Content-Type": "application/json",
+					Accept: "application/json",
```

Comment:
We don't use getJsonHeaders() here because this specific request doesn't require an access token for authorization, which getJsonHeaders() is designed to handle when available. In this case, manually specifying the necessary headers suffices since only the content type and accept type are needed for the OAuth token request.